proposals/2024-01-29_native_histograms_text_format.md

```diff
@@ -0,0 +1,110 @@
+## Native Histograms Text Format
+
+* **Owners:**
+  * Chris Marchbanks [@csmarchbanks](https://github.com/csmarchbanks)
+
+* **Implementation Status:** Not implemented
+
+* **Related Issues and PRs:**
+  * [Native Histogram Support in client_python](https://github.com/prometheus/client_python/issues/918)
+  * [OpenMetrics Protobuf format PR](https://github.com/OpenObservability/OpenMetrics/pull/256)
+
+* **Other docs or links:**
+  * [Design Doc for choosing a proposal](https://docs.google.com/document/d/1qoHf24cKMpa1QHskIjgzyf3oFhIPvacyJj8Tbe6fIrY/edit#heading=h.5sybau7waq2q)
+
+> TL;DR: This design doc is proposing a format for exposing native histograms in the OpenMetrics text format.
+
+## Why
+
+Today it is only possible to export native histograms using the classic (not OpenMetrics) Protocol Buffers (protobuf) scrape format. Many users prefer the text format, and some client libraries, such as the Python client, want to avoid adding a dependency on protobuf. Prometheus content negotiation prefers OpenMetrics to the classic Prometheus text based format, therefore to support native histograms in Prometheus the OpenMetrics text format will also need to support native histograms.
+
+There is already an open pull request (see Related Issues and PRs above) to add support for native histograms to OpenMetrics, and during a [dev summit in 2022](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#bookmark=id.c3e7ur6rn5d2) there was consensus we would continue to support the text format for new features as well. Including native histograms as part of the text format shows commitment to that consensus.
+
+### Pitfalls of the current solution
+
+Prometheus client libraries such as Python do not want to require a dependency on protobuf in order to expose native histograms, and in some languages protobuf is painful to use. Gating native histograms only to clients/users willing to use protobuf hurts adoption of native histograms, therefore, we would like a way to represent a native histogram in the text based format.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Support native histograms in the text format
+* (Secondary) Encode/decode efficiency
+* (Secondary) Ease of implementation for client libraries
+* (Secondary) Human readibility of the format
+
+Note that the goals of efficiency and human readability are commonly at odds with each other.
+
+### Audience
+
+Client library maintainers, OpenMetrics, and Prometheus scrape maintainers.
+
+## Non-Goals
```

Comment:
Also a non goal (I guess): Created timestamp on native histograms

Reply:
I am neutral on support for created timestamps, seems unlikely too many people will use them, but:
```
my_histogram {...value...} 1717524294.000
```
would work with this proposal.

Reply:
I wouldn't change how created timestamps are represented in OM in this proposal. We sure want to change it, but it would probably be better to change created timestamps in all metric types at once instead of changing for native histograms and others later.

Reply:
FYI: [here is how created timestamps are represented for histograms in OM today](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#histogram-1:~:text=in%20any%20order.-,Histogram,appear%20in%20a%20Sample%20with%20a%20MetricName%20with%20the%20suffix%20%22_count%22.,-Buckets%20MUST%20be).

For native histograms I think we could to the same for now? Just add a new line, repeating the metric name and using the suffix `_created`

Reply:
Ahh, missed the "Created" part of the original comment. I agree that we will not change how created timestamps are represented in OpenMetrics today. We will continue to add an additional `_created` when native histograms are in use.

Reply:
Do you mind adding this to non goal? e.g. `* Changing created timestamp semantics (it will be a separate series as in OM 1.0)`

Reply:
Done!